### PR TITLE
Improve regular expression error messages

### PR DESCRIPTION
**Pull request by anthony-chang**
Closes https://github.com/NVIDIA/spark-rapids/issues/4537

This PR improves the error messages when throwing `RegexUnsupportedExceptions` by attempting to isolate what exactly in the pattern is not supported. I also added a `position` field to `RegexAST` and recorded the index of each node in the AST within the regular expression pattern in the `parse` stage, so that this can be printed in errors thrown in the `rewrite` stage. All `RegexUnsupportedExceptions` must specify the index now.

There should be no changes to functionality.

### CODE REVIEW

File: tests/src/test/scala/com/nvidia/spark/rapids/RegularExpressionTranspilerSuite.scala [hunk 1]
```diff
@@ -221,28 +228,31 @@ class RegularExpressionTranspilerSuite extends FunSuite with Arm {
...
+        "sequences that only contain '^' or '$' are not supported")
     )
   }
 
   test("string anchor \\Z fall back to CPU in some repetitions") {
     val patterns = Seq(raw"a(\Z)*", raw"a(\Z){2,}")
     patterns.foreach(pattern =>
-      assertUnsupported(pattern, RegexFindMode, "nothing to repeat")
+      assertUnsupported(pattern, RegexFindMode, 
+        "cuDF does not support reptition of group containing: \\Z")
```
Comment (andygrove) [inline tests/src/test/scala/com/nvidia/spark/rapids/RegularExpressionTranspilerSuite.scala:248]:
```suggestion
        "cuDF does not support repetition of group containing: \\Z")
```

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala [hunk 2]
```diff
@@ -1269,35 +1297,46 @@ class CudfRegexTranspiler(mode: RegexMode) {
             // NOTE: (\A)* can be transpiled to (\A)?
             // however, (\A)? is not supported in libcudf yet
             case _ =>
-              throw new RegexUnsupportedException(nothingToRepeat)
+              val unsupportedTerm = isSupportedRepetitionBase(term)._2.get
+              throw new RegexUnsupportedException(
+                s"cuDF does not support reptition of group containing: " +
```
Comment (andygrove) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala:1302]:
```suggestion
                s"cuDF does not support repetition of group containing: " +
```

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala [hunk 3]
```diff
@@ -143,13 +148,15 @@ class RegexParser(pattern: String) {
         parseEscapedCharacter()
       case '\u0000' =>
         throw new RegexUnsupportedException(
-          "cuDF does not support null characters in regular expressions", Some(pos))
+          "cuDF does not support null characters in regular expressions", Some(pos-1))
       case '*' | '+' | '?' =>
         throw new RegexUnsupportedException(
-          "base expression cannot start with quantifier", Some(pos))
+          "base expression cannot start with quantifier", Some(pos-1))
```
Comment (andygrove) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala:154]: nit: Error messages are currently inconsistent with respect to the capitalization of the first word. I suggest starting them all with upper-case with the exception of messages starting with "cuDF ..."

```suggestion
          "Base expression cannot start with quantifier", Some(pos-1))
```

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala [hunk 4]
```diff
@@ -263,17 +271,15 @@ class RegexParser(pattern: String) {
                   characterClass.appendRange(nextChar, RegexChar(end))
                 case _ =>
                   throw new RegexUnsupportedException(
-                    "unexpected EOF while parsing character range",
-                    Some(pos))
+                    "unexpected EOF while parsing character range", Some(pos))
```
Comment (andygrove) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala:274]:
```suggestion
                    "Unexpected EOF while parsing character range", Some(pos))
```

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala [hunk 5]
```diff
@@ -956,7 +965,7 @@ class CudfRegexTranspiler(mode: RegexMode) {
                 RegexChar('$')))
             case Some(RegexEscaped('b')) | Some(RegexEscaped('B')) =>
               throw new RegexUnsupportedException(
-                      "regex sequences with \\b or \\B not supported around $")
+                      "regex sequences with \\b or \\B not supported around $", regex.position)
```
Comment (andygrove) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala:968]:
```suggestion
                      "Regex sequences with \\b or \\B not supported around $", regex.position)
```

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala [hunk 6]
```diff
@@ -1032,14 +1042,17 @@ class CudfRegexTranspiler(mode: RegexMode) {
           }
         case 'b' | 'B' if mode == RegexSplitMode =>
           // see https://github.com/NVIDIA/spark-rapids/issues/5478
-          throw new RegexUnsupportedException("word boundaries are not supported in split mode")
+          throw new RegexUnsupportedException(
+            "word boundaries are not supported in split mode", regex.position)
```
Comment (andygrove) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala:1046]:
```suggestion
            "Word boundaries are not supported in split mode", regex.position)
```

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala [hunk 7]
```diff
@@ -1032,14 +1042,17 @@ class CudfRegexTranspiler(mode: RegexMode) {
...
         case 'A' if mode == RegexSplitMode =>
-          throw new RegexUnsupportedException("string anchor \\A is not supported in split mode")
+          throw new RegexUnsupportedException(
+            "string anchor \\A is not supported in split mode", regex.position)
```
Comment (andygrove) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala:1049]:
```suggestion
            "String anchor \\A is not supported in split mode", regex.position)
```

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala [hunk 8]
```diff
@@ -1032,14 +1042,17 @@ class CudfRegexTranspiler(mode: RegexMode) {
...
         case 'Z' if mode == RegexSplitMode =>
           throw new RegexUnsupportedException(
-              "string anchor \\Z is not supported in split or replace mode")
+            "string anchor \\Z is not supported in split or replace mode", regex.position)
```
Comment (andygrove) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala:1052]:
```suggestion
            "String anchor \\Z is not supported in split or replace mode", regex.position)
```

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala [hunk 9]
```diff
@@ -1032,14 +1042,17 @@ class CudfRegexTranspiler(mode: RegexMode) {
...
         case 'z' if mode == RegexSplitMode =>
-          throw new RegexUnsupportedException("string anchor \\z is not supported in split mode")
+          throw new RegexUnsupportedException(
+            "string anchor \\z is not supported in split mode", regex.position)
```
Comment (andygrove) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala:1055]:
```suggestion
            "String anchor \\z is not supported in split mode", regex.position)
```

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala [hunk 10]
```diff
@@ -1135,26 +1150,34 @@ class CudfRegexTranspiler(mode: RegexMode) {
...
+          // example: "|b"
+          throw new RegexUnsupportedException("choice with one empty side not supported", 
+            parts.head.position)
         }
-        if (isRegexChar(parts.head, '|') || isRegexChar(parts.last, '|')) {
-          // examples: "a|", "|b"
-          throw new RegexUnsupportedException(nothingToRepeat)
+        if (isRegexChar(parts.last, '|')) {
+          // example: "a|"
+          throw new RegexUnsupportedException("choice with one empty side not supported",
```
Comment (andygrove) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala:1166]:
```suggestion
          throw new RegexUnsupportedException("Choice with one empty side not supported", 
```

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala [hunk 11]
```diff
@@ -1135,26 +1150,34 @@ class CudfRegexTranspiler(mode: RegexMode) {
...
+            parts.last.position)
         }
         if (isRegexChar(parts.head, '{')) {
           // example: "{"
           // cuDF would treat this as a quantifier even though in this
           // context (being at the start of a sequence) it is not quantifying anything
           // note that we could choose to escape this in the transpiler rather than
           // falling back to CPU
-          throw new RegexUnsupportedException(nothingToRepeat)
+          throw new RegexUnsupportedException("token preceding '{' is not quantifiable",
```
Comment (andygrove) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala:1175]:
```suggestion
          throw new RegexUnsupportedException("Token preceding '{' is not quantifiable", 
```

**Comment by anthony-chang** [comment]
build

**Review by NVnavkumar — APPROVED** [review]

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala [hunk 12]
```diff
@@ -1240,24 +1264,28 @@ class CudfRegexTranspiler(mode: RegexMode) {
...
 
         case (_, QuantifierFixedLength(0)) if mode != RegexFindMode =>
           throw new RegexUnsupportedException(
-            "regex_replace and regex_split on GPU do not support repetition with {0}")
+            "regex_replace and regex_split on GPU do not support repetition with {0}",
+            quantifier.position)
 
         case (RegexGroup(capture, term), SimpleQuantifier(ch))
-            if "+*".contains(ch) && !isSupportedRepetitionBase(term) =>
+            if "+*".contains(ch) && !(isSupportedRepetitionBase(term)._1) =>
```
[the remark this answers is not in the document]
Comment (anthony-chang) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala:1299]: I'll do this in a separate PR

**Pull request merged**